File: java/v1/src/test/java/com/api/v1/BookRegistrationTest.java
package com.api.v1;

import com.api.v1.dtos.requests.BookRegistrationRequestDto;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.reactive.server.WebTestClient;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class BookRegistrationTest {

    @Autowired
    private WebTestClient webTestClient;

    @Test
    void testSuccessfulBookRegistration() {

        var registrationRequest = new BookRegistrationRequestDto(
                "Harry Potter",
                "",
                "123456789012",
                "J.K. Rowling",
                "juvenile fantasy",
                300,
                1,
                120.0,
                "Penguin Inc.",
                1997
        );

        webTestClient
                .post()
                .uri("api/v1/books")
                .bodyValue(registrationRequest)
                .exchange()
                .expectStatus()
                .is2xxSuccessful();

    }

    @Test
    void testUnsuccessfulBookRegistration() {

        var registrationRequest = new BookRegistrationRequestDto(
                "Harry Potter",
                "",
                "123456789012",
                "J.K. Rowling",
                "juvenile fantasy",
                300,
                1,
                120.0,
                "Penguin Inc.",
                1997
        );

        webTestClient
                .post()
                .uri("api/v1/books")
                .bodyValue(registrationRequest)
                .exchange()
                .expectStatus()
                .is5xxServerError();

    }

}

File: kotlin/v2/src/test/kotlin/com/api/v2/BookRegistrationTest.kt
package com.api.v2

import com.api.v2.book.dtos.BookRegistrationRequestDto
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.reactive.server.WebTestClient

@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
private class BookRegistrationTest {

    @Autowired
    lateinit var webTestClient: WebTestClient

    val requestDto = BookRegistrationRequestDto(
        "Harry Porter",
        "The First Book",
        "1234567890123",
        "J.K. Rowling",
        "juvenile fantasy",
        "HP Collins",
        300,
        1,
        1997,
        100.0
    )

    @Test
    @Order(1)
    fun testSuccessfulRegistration() {
        webTestClient
            .post()
            .uri("api/v2/books")
            .bodyValue(requestDto)
            .exchange()
            .expectStatus()
            .is2xxSuccessful()
    }

    @Test
    fun testUnsuccessfulRegistration() {
        webTestClient
            .post()
            .uri("api/v2/books")
            .bodyValue(requestDto)
            .exchange()
            .expectStatus()
            .is5xxServerError()
    }

}